pallets/stablewap/src/lib.rs

```diff
@@ -0,0 +1,707 @@
+			amount_in: Balance,
+			amount_out: Balance,
+			fee: Balance,
+		},
+	}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq))]
+	pub enum Error<T> {
+		/// Creating a pool with same assets is not allowed.
+		SameAssets,
+
+		/// A pool with given assets does not exist.
+		PoolNotFound,
+
+		/// A pool with given assets already exists.
+		PoolExists,
+
+		/// Asset is not in the pool.
+		AssetNotInPool,
+
+		/// One or more assets are not registered in AssetRegistry
+		AssetNotRegistered,
+
+		/// Invalid asset amount provided. Amount must be greater than zero.
+		InvalidAssetAmount,
+
+		/// Balance of an asset is nto sufficient to perform a trade.
+		InsufficientBalance,
+
+		/// Balance of an share asset is nto sufficient to withdraw liquidity.
+		InsufficientShares,
+
+		/// Liquidity has not reached the required minimum.
+		InsufficientLiquidity,
+
+		/// Insufficient liquidity left in the pool after withdrawal.
+		InsufficientLiquidityRemaining,
+
+		/// Amount is less than min trading limit.
+		InsufficientTradingAmount,
+
+		/// Minimum limit has not been reached during trade.
+		BuyLimitNotReached,
+
+		/// Maximum limit has been exceeded during trade.
+		SellLimitExceeded,
+
+		/// Initial liquidity of asset must be > 0.
+		InvalidInitialLiquidity,
+
+		/// Account balance is too low.
+		BalanceTooLow,
+
+		/// Amplification is outside specified range.
+		InvalidAmplification,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a 2-asset pool with initial liquidity for both assets.
+		///
+		/// Both assets must be correctly registered in `T::AssetRegistry`
+		///
+		/// Initial liquidity must be > 0.
+		///
+		/// Origin is given corresponding amount of shares.
+		///
+		/// Parameters:
+		/// - `origin`: Must be T::CreatePoolOrigin
+		/// - `assets`: Asset ids tuple
+		/// - `initial_liquidity`: Corresponding initial liquidity of `assets`
+		/// - `amplification`: Pool amplification
+		/// - `fee`: trade fee to be applied in sell/buy trades
+		///
+		/// Emits `PoolCreated` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::create_pool())]
+		#[transactional]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			assets: (T::AssetId, T::AssetId),
+			initial_liquidity: (Balance, Balance),
+			amplification: u32,
+			fee: Permill,
+		) -> DispatchResult {
+			let who = T::CreatePoolOrigin::ensure_origin(origin)?;
+
+			let pool_assets: PoolAssets<T::AssetId> = assets.into();
+
+			ensure!(pool_assets.is_valid(), Error::<T>::SameAssets);
+
+			ensure!(
+				T::AmplificationRange::get().contains(&amplification),
+				Error::<T>::InvalidAmplification
+			);
+
+			for asset in (&pool_assets).into_iter() {
+				ensure!(T::AssetRegistry::exists(asset), Error::<T>::AssetNotRegistered);
+			}
+
+			ensure!(
+				initial_liquidity.0 > Balance::zero(),
+				Error::<T>::InvalidInitialLiquidity
+			);
+			ensure!(
+				initial_liquidity.1 > Balance::zero(),
+				Error::<T>::InvalidInitialLiquidity
+			);
```

Comment:
Why not use the `valid` (or according to my suggestion `is_valid`;-) ) method on `AssetAmounts`?
You create an `AssetAmounts` below AFAICT